Editing the mention should trigger the typeahead again

aniyati opened the pull request:
Editing the mention should trigger the typeahead again

- On removal of whitespace between control character and typed in string, it should trigger mention again
- Updated state machine, to update the flow to trigger mention

Note: The base branch is aniyati/Hakawai/PunctuationMention to exclude the changes from that branch from the diff.

aniyati committed:
Editing the mention should trigger the typeahead again

- On removal of whitespace between control character and typed in string, it should trigger mention again
- Updated state machine, to update the flow to trigger mention

bkoatz commented on Hakawai/Mentions/HKWMentionsCreationStateMachine.m:
```diff
@@ -164,7 +164,10 @@ - (void)characterTyped:(unichar)c {
     }
-    if ([self.stringBuffer length] == 0 && isWhitespace) {
+    // When whitespace is typed during mention creation state and previous char is control character
+    // then metion creation should to stalled. e.g "@@ " will stop mention current creation.
```
should be ended. e.g "@@ " will stop mention current creation.

aniyati replied:
Thanks for the input. Corrected code comment.

bkoatz commented on Hakawai/Mentions/HKWMentionsStartDetectionStateMachine.m:
```diff
@@ -169,15 +172,19 @@ - (void)characterTyped:(unichar)c asInsertedCharacter:(BOOL)inserted previousCha
-                        // Empty string buffer when previous character is a punctuation
-                        self.stringBuffer = [@"" mutableCopy];
+                    NSString *keyString;
+                    if (queryMentionForText) {
```
what is happening here

aniyati replied:
`queryMentionForText` (now renamed to `nextString`) holds the word succeeding the control character. When a control character is inserted (scenario: _user typed abc and inserted "@" before this string_), we are setting the string buffer to `queryMentionForText` (`nextString`) to look for a matching mention.

bkoatz commented on Hakawai/Mentions/HKWMentionsPlugin.m:
```diff
@@ -844,7 +844,10 @@ - (BOOL)advanceStateForCharacterInsertion:(unichar)newChar
+
+            // When control character is inserted before existing word in text view, then query mention with that word.
+            NSString *textAfterControlCharacter = [HKWMentionsStartDetectionStateMachine wordAfterLocation:location text:parentTextView.text];
+            [self.startDetectionStateMachine characterTyped:newChar asInsertedCharacter:NO previousCharacter:precedingChar queryMentionForText:textAfterControlCharacter];
```
why do we need both newChar and this other information?

aniyati replied:
`textAfterControlCharacter` contains word succeeding the control character. This is needed to start a new mention with matching text.

bkoatz commented on Hakawai/Mentions/HKWMentionsStartDetectionStateMachine.m:
```diff
@@ -202,22 +209,46 @@ - (void)characterTyped:(unichar)c asInsertedCharacter:(BOOL)inserted previousCha
+        currentCharacterType = CharacterTypeSeparator;
+    } else {
+        // Check first for control character because punctuation character can be control character.
+        // e.g "@" and "#" are punctuation char and might be control character.
```
chars* characters*

aniyati replied:
Updated here and elsewhere :)

bkoatz commented on Hakawai/Mentions/HKWMentionsStartDetectionStateMachine.m:
```diff
@@ -202,22 +209,46 @@ - (void)characterTyped:(unichar)c asInsertedCharacter:(BOOL)inserted previousCha
 
+    // Determine the type of the character
+    enum CharacterType currentCharacterType = CharacterTypeNormal;
+    if ([HKWMentionsStartDetectionStateMachine.whitespaceSet characterIsMember:precedingChar] || precedingChar == 0) {
```
there should be a function where you pass in a char and ire returns the chartype

bkoatz commented on Hakawai/Mentions/HKWMentionsStartDetectionStateMachine.m:
```diff
@@ -202,22 +209,46 @@ - (void)characterTyped:(unichar)c asInsertedCharacter:(BOOL)inserted previousCha
+            // If user deletes white-space between control character and word, then query mention with word next to whitepace.
+            if (currentCharacterType == CharacterTypeControlCharacter && deletedCharacterType == CharacterTypeSeparator) {
+                self.state = HKWMentionsStartDetectionStateCreatingMention;
+                NSString *const keyword = [HKWMentionsStartDetectionStateMachine wordAfterLocation:location + 1 text:textViewText];
```
if you're doing +1s and -1s you should be absolutely sure you're within the index bounds

bkoatz commented on Hakawai/Mentions/HKWMentionsStartDetectionStateMachine.m:
```diff
@@ -298,6 +330,20 @@ - (void)mentionCreationResumed {
+    NSMutableString *const word = [[NSMutableString alloc] init];
+    for(NSUInteger i = location; i < text.length ; i++) {
+        charString =  [text substringWithRange:NSMakeRange(i, 1)];
+        if([text isEqualToString:@" "]) {
```
what does this do

aniyati replied:
When a control character ("@" or "#") is inserted before an already existing word, we would need that word, to begin/trigger a mention with or replace the word with the selected mention. This method returns the characters after a certain location till the next delimiter (whitespace/newline) as a word
Updated logic to use `whitespaceAndNewLine` character set instead hard code checks.

bkoatz commented on Hakawai/Mentions/HKWMentionsPlugin.m:
```diff
@@ -844,7 +844,10 @@ - (BOOL)advanceStateForCharacterInsertion:(unichar)newChar
-            [self.startDetectionStateMachine characterTyped:newChar asInsertedCharacter:NO previousCharacter:precedingChar];
+
+            // When control character is inserted before existing word in text view, then query mention with that word.
+            NSString *textAfterControlCharacter = [HKWMentionsStartDetectionStateMachine wordAfterLocation:location text:parentTextView.text];
```
doesn't using this break down some core elements of the hakawai state machine? like if we just look up the word in each moment, rather than using the set location and prev/next chars, etc. that are being set throughout the code? Let's talk on slack

aniyati replied:
Here we want to achieve functionality where mentioned should be detected even after inserting @ after some name.
We will ping you over slack

bkoatz replied:
you don't know where this is a control character yet, so update the comment and call this wordFollowingTypedCharacter

aniyati replied:
Updated comment and renamed `textAfterControlCharacter`.

aniyati committed:
When a control character is inserted before an already existing word, we would need the word to begin a mention with or replace the word with the selected mention

Commit:
Merge pull request #132 from aniyati/PunctuationMention

Adds support to trigger mention after punctuation marks like brackets, braces

bkoatz commented on Hakawai/Mentions/HKWMentionsCreationStateMachine.m:
```diff
@@ -164,7 +164,10 @@ - (void)characterTyped:(unichar)c {
     }
-    if ([self.stringBuffer length] == 0 && isWhitespace) {
+    // When whitespace is typed during mention creation state and previous character is control character
+    // then mention creation should be ended. e.g "@@ " will stop mention current creation.
```
sorry to change my mind, but this is slightly cleaner:

should end. e.g "@@ " will stop current mention creation.

bkoatz commented on Hakawai/Mentions/HKWMentionsCreationStateMachine.m:
```diff
@@ -164,7 +164,10 @@ - (void)characterTyped:(unichar)c {
+    // When whitespace is typed during mention creation state and previous character is control character
+    // then mention creation should be ended. e.g "@@ " will stop mention current creation.
+    const BOOL shouldCreateNewMentionState = previousCharacterIsControl && isWhitespace;
+    if (([self.stringBuffer length] == 0 && isWhitespace) || shouldCreateNewMentionState) {
```
this should be just "([self.stringBuffer length] == 0 || previousCharacterIsControl) && isWhitespace"

also can't previousCharacterIsControl be just replaced with whether the "stack" has just a "@" in it? Cuz wouldn't that be the state of the stack here?

aniyati replied:
Modified if logic. 
The string buffer contains the mention string and checking if it contains "@" would not fulfill our requirements. Please correct us if we misunderstood something.

bkoatz replied:
then isn't stringBuffer lenght and previousCharacterIsControl the same? You can talk to me on slack :)

aniyati replied:
We have noticed that mentions does not stop even after a whitespace is encountered and continues to search for matching mentions till no mention can be found. These changes were for the case of "@@ "(Note the space at 3rd position), for which the length of string buffer is not zero but the previous character is control. If the whitespace after "@@"(First two characters) is deleted, it has to trigger mentions as its previous character is control character and character before previous character is punctuation.
We have removed these changes from this PR and can create another PR for this if required.

bkoatz commented on Hakawai/Mentions/HKWMentionsCreationStateMachine.m:
```diff
@@ -276,6 +279,13 @@ - (void)stringDeleted:(NSString *)deleteString {
             // User not creating a mention right now
             return;
         case HKWMentionsCreationStateCreatingMention:
+            if (isControlCharacterDeleted) {
```
can't that just be determined by if deleteString is a "@"?

bkoatz replied:
also isn't this case handled by 

else if ([deleteString length] > [self.stringBuffer length]) {
                // Delete will completely clear out the string buffer
                [self cursorMoved];
                return;
            }

aniyati replied:
> can't that just be determined by if deleteString is a "@"?

As "@" can be a part of a normal text as well, we want to know if a mention triggering "@" was deleted. We are passing `isControlCharacterDeleted` flag by checking if the deleted character is a control character that triggers mentions.

aniyati replied:
> also isn't this case handled by
> 
> else if ([deleteString length] > [self.stringBuffer length]) {
> // Delete will completely clear out the string buffer
> [self cursorMoved];
> return;
> }

For the case when an "@" is inserted before an alphanumeric word and a mention starts, if the "@" is deleted soon after mention is triggered, `deleteString`'s length would be 1 (as we are deleting "@") whereas `stringBuffer`'s length would be the length of the string for which mention was triggered which can be greater than 1 for which this condition fails.

bkoatz replied:
could this be just if "@" is the last character in the string buffer? Cuz that's the only time a control character would be deleted in mentions mode without ending the mention - so if that's NOT true, we can end the mention

bkoatz commented on Hakawai/Mentions/HKWMentionsPlugin.m:
```diff
@@ -858,7 +864,7 @@ - (BOOL)advanceStateForCharacterInsertion:(unichar)newChar
             // Inform the mentions creation state machine that a character was typed. Do not allow the double space to
             //  period auto-substitution while the user is creating a mention.
-            [self.creationStateMachine characterTyped:newChar];
+            [self.creationStateMachine characterTyped:newChar previousCharacterIsControl:[self.controlCharacterSet characterIsMember:precedingChar]];
```
yeah this seems like it can just happen in the function itself without need a new argument

aniyati replied:
As we do not have previous character or its type in the function, we will have to pass the value.

bkoatz commented on Hakawai/Mentions/HKWMentionsPlugin.m:
```diff
@@ -1238,11 +1258,22 @@ - (BOOL)advanceStateForStringDeletionAtRange:(NSRange)range
-        case HKWMentionsStartDetectionStateCreatingMention:
-            [self.creationStateMachine stringDeleted:deletedString];
+        case HKWMentionsStartDetectionStateCreatingMention: {
+            const BOOL precedingCharacterIsWhitepace = [[NSCharacterSet whitespaceAndNewlineCharacterSet] characterIsMember:precedingCharacter];
```
why does preceding character being whitespace matter here

aniyati replied:
As we want to know if control character that triggers mention is deleted, we were checking for white space before deleted string. We have now changed white space check to separator check.

bkoatz commented on Hakawai/Mentions/HKWMentionsPlugin.m:
```diff
@@ -1781,7 +1812,12 @@ - (void)createMention:(HKWMentionsAttribute *)mention startingLocation:(NSUInteg
+    // we want to replace word after control character with mention text.
+    // e.g "hey @|john" will be replaced as "hey John Doe". '|' indicates cursor.
+    NSString *const wordAfterCurrentLocation = [HKWMentionsStartDetectionStateMachine wordAfterLocation:currentLocation text:parentTextView.text];
+    NSRange rangeToTransform = NSMakeRange(location, currentLocation + wordAfterCurrentLocation.length - location);
```
should currentLocation just be updated when you type an "@" in the first place? Rather than editing it at the end here?

aniyati replied:
We are not updating current location here. The current location would be the same. We are trying to get the length of range to transfer - which would be the sum of difference between the current location and starting locations, and length of word after current location, as we want to replace the whole word succeeding the control character.

bkoatz replied:
couldn't we just use mention selected range? cuz there are variables for that that should be up to date. so this range could just be from current mention start location, with length of current mention length

if you have trouble finding this u can let me know

bkoatz commented on Hakawai/Mentions/HKWMentionsPlugin.m:
```diff
@@ -987,7 +995,17 @@ - (BOOL)advanceStateForCharacterDeletion:(unichar)precedingChar
+
+            // This check is used to stop mention creation when control character is deleted.
+            // Also check if second previous character is control, it YES then don't stop mention creation.
+            BOOL secondPreviousCaracterIsControl = NO;
```
can you link me to the specific mentions bug JIRA that product has given for this to fix? (in other words: the bug of "mentions should remain if two mentions characters were typed in a row")

also what if someone typed 3 "@" characters in a row?

aniyati replied:
Currently, mentions is triggered when a control character is typed that satisfies all the conditions of creating a mention. Once mentions is triggered, search starts regardless of what is typed after control character.
If 3 "@" are typed in a row, mentions will be triggered for 2 "@" ie., for "@@"

bkoatz commented on Hakawai/Mentions/HKWMentionsStartDetectionStateMachine.m:
```diff
@@ -112,37 +123,14 @@ - (void)validStringInserted:(NSString *)string
+- (void)characterTyped:(unichar)c
+   asInsertedCharacter:(BOOL)inserted
+     previousCharacter:(unichar)previousCharacter
+            nextString:(nullable NSString *)nextString {
```
can you call this 

wordFollowingTypedCharacter:

and document this function?

aniyati replied:
Renamed `nextString` to `wordFollowingTypedCharacter` and documented the function.

bkoatz commented on Hakawai/Mentions/HKWMentionsStartDetectionStateMachine.m:
```diff
@@ -202,22 +191,36 @@ - (void)characterTyped:(unichar)c asInsertedCharacter:(BOOL)inserted previousCha
+            BOOL shouldCreateMention = NO;
+            if (location > 1 && textViewText.length > location - 2) {
+                const unichar characterBeforePrecedingChar = [textViewText characterAtIndex:location-2];
+                shouldCreateMention = [HKWMentionsStartDetectionStateMachine.separatorSet characterIsMember:characterBeforePrecedingChar];
```
wouldn't this also be true if the mentions character is the first character in the screen? eg, if precedingChar is the first char?

bkoatz replied:
also this should more accurately be called "isCharacterBeforePrecedingCharASeparator" or something like that, because we don't know if current character type is a control char or not

bkoatz replied:
also also what if the characterbeforeprecedingchar is a control char? what happens then?

aniyati replied:
> wouldn't this also be true if the mentions character is the first character in the screen? eg, if precedingChar is the first char?

Before checking for `characterBeforePrecedingChar`, we are checking for location greater than 1. This would not include the case when preceding character is the first character.
 
We have made changes to set `canCreateMention` (previously `shouldCreateMention`) to true when preceding character is the first character and is a control character.

aniyati replied:
> also this should more accurately be called "isCharacterBeforePrecedingCharASeparator" or something like that, because we don't know if current character type is a control char or not

We have now renamed `shouldCreateMention` as `canCreateMention` as we have now included the check of preceding character type and it tells us if it is possible to create a mention.

aniyati replied:
> also also what if the characterbeforeprecedingchar is a control char? what happens then?

Right now, we are checking for separators (separator set that includes control characters) for character before preceding character and mentions will trigger when this character is a separator(regardless of it being a control character). Please let us know if this behavior is not intended.

bkoatz replied:
so a control character is a separator also? is that within the spec of this change? I thought this was just a fix to allow @ mentions after things like parentheses?

Talk to me on slack :)

aniyati replied:
As control characters can overlap with separators, we are treating control characters as separators wherever they do not fulfill conditions of creating a mention.

bkoatz commented on Hakawai/Mentions/HKWMentionsStartDetectionStateMachine.m:
```diff
@@ -202,22 +191,36 @@ - (void)characterTyped:(unichar)c asInsertedCharacter:(BOOL)inserted previousCha
+                textViewText:(nonnull NSString *)textViewText {
+    // Determine the character types
+    enum CharacterType deletedCharacterType = [self characterTypeOfCharacter:deletedChar];
+    enum CharacterType currentCharacterType = [self characterTypeOfCharacter:precedingChar];
```
rename this precedingCharacterType

aniyati replied:
Renamed `currentCharacterType` to `precedingCharacterType`.

bkoatz commented on Hakawai/Mentions/HKWMentionsStartDetectionStateMachine.m:
```diff
@@ -202,22 +191,36 @@ - (void)characterTyped:(unichar)c asInsertedCharacter:(BOOL)inserted previousCha
+                const unichar characterBeforePrecedingChar = [textViewText characterAtIndex:location-2];
+                shouldCreateMention = [HKWMentionsStartDetectionStateMachine.separatorSet characterIsMember:characterBeforePrecedingChar];
+            }
+            // If user deletes white-space between control character and word, then query mention with word next to whitepace.
```
not just white-space, but separators and control characters right?

aniyati replied:
Corrected code comment.

bkoatz commented on Hakawai/Mentions/HKWMentionsStartDetectionStateMachine.m:
```diff
@@ -202,22 +191,36 @@ - (void)characterTyped:(unichar)c asInsertedCharacter:(BOOL)inserted previousCha
+            }
+            // If user deletes white-space between control character and word, then query mention with word next to whitepace.
+            if (currentCharacterType == CharacterTypeControlCharacter
+                && (deletedCharacterType == CharacterTypeSeparator || deletedCharacterType == CharacterTypeControlCharacter)
```
isn't the controlCharacter part of this logic handled by "secondPreviousCaracterIsControl" logic above?

aniyati replied:
We were only checking for `characterBeforePrecedingChar`'s type to set `shouldCreateMention`.  We have now included `precedingCharacterType == CharacterTypeControlCharacter` check in `canCreateMention` (previously `shouldCreateMention`) and removed the same check from here.

bkoatz commented on Hakawai/Mentions/HKWMentionsStartDetectionStateMachine.m:
```diff
@@ -298,6 +295,46 @@ - (void)mentionCreationResumed {
+
+#pragma mark - Private helper method
+
+- (CharacterType)characterTypeOfCharacter:(unichar)aCharacter {
```
great work with this

bkoatz reviewed (changes requested):
this is a really hard fix. would take me several days, and I know this library well. so don't get discouraged by number of comments!

also: have you considered interaction of multiple different control char types? like # + @? does this lead to expected behavior? Should we talk to product about this?

aniyati committed:
Fixed review comments.

bkoatz committed:
Set up lix for korean mentions fix

Commit:
Merge pull request #136 from bkoatz/master

Set up lix for korean mentions fix

bkoatz committed:
Implement korean mentions fix

Move logic during mentions from uitextviewdelegate to NSTextStorageDelegate, because textstoragedelegate actually offers us events when korean text is updated while textviewdelegate does not

This holistic solution fixes korean and maintains functionality for all other languages

it is lixed and only applies when creating or starting a mention

bkoatz committed:
Implement korean mentions fix

Move logic during mentions from uitextviewdelegate to NSTextStorageDelegate, because textstoragedelegate actually offers us events when korean text is updated while textviewdelegate does not

This holistic solution fixes korean and maintains functionality for all other languages

it is lixed and only applies when creating or starting a mention

bkoatz committed:
Implement korean mentions fix

Move logic during mentions from uitextviewdelegate to NSTextStorageDelegate, because textstoragedelegate actually offers us events when korean text is updated while textviewdelegate does not

This holistic solution fixes korean and maintains functionality for all other languages

it is lixed and only applies when creating or starting a mention

aniyati committed:
Fixed review comments

aniyati commented:
> this is a really hard fix. would take me several days, and I know this library well. so don't get discouraged by number of comments!
> 
> also: have you considered interaction of multiple different control char types? like # + @? does this lead to expected behavior? Should we talk to product about this?

Commit:
Merge branch 'master' into master

Commit:
Merge pull request #137 from bkoatz/master

Implement korean mentions fix

bkoatz committed:
Release 7.1.3

Commit:
Merge branch 'master' into master

mrsmiley009 committed:
Merge pull request #139 from bkoatz/master

Release 7.1.3

bkoatz commented on Hakawai/Mentions/HKWMentionsPlugin.m:
```diff
@@ -1238,11 +1256,23 @@ - (BOOL)advanceStateForStringDeletionAtRange:(NSRange)range
-        case HKWMentionsStartDetectionStateCreatingMention:
-            [self.creationStateMachine stringDeleted:deletedString];
+        case HKWMentionsStartDetectionStateCreatingMention: {
+            const BOOL precedingCharacterIsSeparator = [[NSCharacterSet whitespaceAndNewlineCharacterSet] characterIsMember:precedingCharacter]
```
could you have an isSeparator function?

bkoatz commented on Hakawai/Mentions/HKWMentionsStartDetectionStateMachine.m:
```diff
@@ -253,21 +258,14 @@ - (void)deleteTypedCharacter:(unichar)deletedChar withCharacterNowPrecedingCurso
             self.charactersSinceLastWhitespace = 0;
-            if (currentCharacterType == CharacterTypeSeparator) {
+            if (currentCharacterType == CharacterTypeSeparator || currentCharacterType == CharacterTypeControlCharacter) {
                 // The user moved the cursor to the beginning of the text region, or right after a newline or whitespace or punctuation
```
update

aniyati replied:
From what we could understand of the comment, this condition to set state as quiescent ready is the current character has to be a punctuation character or whitespace. For the case of '@', it is a control character as well as a punctuation character and `characterTypeOfCharacter` returns control character type and not separator.
To cover both control characters and punctuations, we have added such check. 
Please let us know if we misunderstood the comment.

bkoatz replied:
no i just meant update this comment: "right after a newline or whitespace or punctuation" to include talking about cntrl character

bkoatz commented on Hakawai/Mentions/HKWMentionsStartDetectionStateMachine.m:
```diff
@@ -253,21 +258,14 @@ - (void)deleteTypedCharacter:(unichar)deletedChar withCharacterNowPrecedingCurso
+            if (currentCharacterType == CharacterTypeSeparator || currentCharacterType == CharacterTypeControlCharacter) {
                 // The user moved the cursor to the beginning of the text region, or right after a newline or whitespace or punctuation
                 //  character. This puts the user in the ready state.
                 self.state = HKWMentionsStartDetectionStateQuiescentReady;
```
is this read or creating mention, if you've moved your cursor in front of a control char

aniyati replied:
We feel it should be `ready` as mentions is not triggered yet. 
We have now added functionality that once user types any character when preceding character is control character, mentions would be triggered.
For example, user has typed "hello @ world|" ("|" represents cursor). When cursor is moved to character location after '@', state would be set to ready, which means it does not trigger mentions yet and when user starts typing something after '@', we need to modify `characterTyped` method to trigger mentions for the typed text if expected.

bkoatz commented on Hakawai/Mentions/HKWMentionsStartDetectionStateMachine.m:
```diff
@@ -232,13 +236,14 @@ - (void)deleteTypedCharacter:(unichar)deletedChar withCharacterNowPrecedingCurso
             //   1. A whitespace character is encountered
             //   2. A NON-whitespace character is encountered and a whitespace character was deleted
             //   3. A punctuation character is encountered
-            if (currentCharacterType == CharacterTypeSeparator
```
why did this change, from current to previous, both here and above

aniyati replied:
As per your previous review comment(https://github.com/aniyati/Hakawai/pull/1#discussion_r347678752) and as the character is before the deleted character, we have renamed `currentCharacterType` as `precedingCharacterType`.

bkoatz commented on Hakawai/Mentions/HKWMentionsStartDetectionStateMachine.m:
```diff
@@ -202,22 +191,37 @@ - (void)characterTyped:(unichar)c asInsertedCharacter:(BOOL)inserted previousCha
+                const unichar characterBeforePrecedingChar = [textViewText characterAtIndex:location - 2];
+                canCreateMention = [HKWMentionsStartDetectionStateMachine.separatorSet characterIsMember:characterBeforePrecedingChar]
+                && precedingCharacterType == CharacterTypeControlCharacter;
+            } else if (location > 0 && precedingCharacterType == CharacterTypeControlCharacter) {
```
why do u need the location check here?

aniyati replied:
The location check was to check only for location 1. We now see that not having location check would not affect behavior for location 0. So we have removed the check.

bkoatz commented on Hakawai/Mentions/HKWMentionsStartDetectionStateMachine.m:
```diff
@@ -202,22 +191,37 @@ - (void)characterTyped:(unichar)c asInsertedCharacter:(BOOL)inserted previousCha
     switch (self.state) {
         case HKWMentionsStartDetectionStateQuiescentReady: {
-            if (currentCharacterType == CharacterTypeNormal) {
+            BOOL canCreateMention = NO;
```
do we wanna do this at the call site for delete, with a function in that class, so we dont have to pass all this extra info in, and just pass in "canCreateMention"

aniyati replied:
If we would want to move the `canCreateMention` logic to caller, we would still have to pass
- location (to begin mentions at that particular location) 
- keyword (to search mentions for that particular word) 

which would not be helping us in reducing the number of parameters to be passed and also as we want to check the character type which can be done better here as we have an enum and helper method in place.

bkoatz commented on Hakawai/Mentions/HKWMentionsStartDetectionStateMachine.m:
```diff
@@ -202,22 +191,37 @@ - (void)characterTyped:(unichar)c asInsertedCharacter:(BOOL)inserted previousCha
 
     switch (self.state) {
         case HKWMentionsStartDetectionStateQuiescentReady: {
-            if (currentCharacterType == CharacterTypeNormal) {
```
can u add comments to this code

aniyati replied:
Added code comment

bkoatz reviewed (changes requested):
have u merged this with the hakawai fixes and made sure this still works?

aniyati committed:
Fixed review comment

bkoatz commented on Hakawai/Mentions/HKWMentionsStartDetectionStateMachine.m:
```diff
@@ -203,12 +203,16 @@ - (void)deleteTypedCharacter:(unichar)deletedChar
+            // Mention can be triggered upon character deletion:
+            // 1. when deleted character location is greater than 1 -> When previous character is a control character and character before previous character is a separator.
+            // 2. when deleted character location is 1 -> If preceding character is a control character
+            // 3. when deleted character location is 0 -> Does not trigger mentions
```
change this whole line to "// (NOTE: When deleted character location is 0, mentions is never triggered)

bkoatz commented on Hakawai/Mentions/HKWMentionsCreationStateMachine.m:
```diff
@@ -270,6 +270,14 @@ - (void)stringDeleted:(NSString *)deleteString isControlCharacterDeleted:(BOOL)i
 
+    BOOL isControlCharacterDeleted = NO;
+    if (deleteString.length == 1
+        &&[deleteString containsString:[NSString stringWithFormat:@"%C", self.explicitSearchControlCharacter]]
```
space after &&

bkoatz commented on Hakawai/Mentions/HKWMentionsCreationStateMachine.m:
```diff
@@ -270,6 +270,14 @@ - (void)stringDeleted:(NSString *)deleteString isControlCharacterDeleted:(BOOL)i
 
     __strong __auto_type delegate = self.delegate;
 
+    BOOL isControlCharacterDeleted = NO;
```
can you add a comment explaining this

bkoatz replied:
When mentions was originally triggered because the whitespace between a control character and a word was deleted, the cursor is next to the control character (like "@|John", where '|' represents the cursor-state). If the user then deletes the control character, the string buffer will not be empty (it will have "John" in it), but mentions has to stop, because control character is deleted. We use the isControlCharacterDeleted flag to decide what to do in this case of control character deletion.

bkoatz commented on Hakawai/Mentions/_HKWMentionsStartDetectionStateMachine.h:
```diff
@@ -92,14 +92,23 @@
  Inform the state machine that a character was deleted by the user from the text view.
  */
-- (void)deleteTypedCharacter:(unichar)deletedChar withCharacterNowPrecedingCursor:(unichar)precedingChar;
+- (void)deleteTypedCharacter:(unichar)deletedChar
```
lets add the param comments here too

aniyati committed:
- Added doc comments

aniyati committed:
Fixed code comment.

dannyighsu committed:
Fix calling an optional protocol method without selector check

Commit:
Add flag to reverse order of mention selection delegate callbacks

Commit:
Merge pull request #142 from mschouest/master

Add flag to reverse order of mention selection delegate callbacks

aniyati committed:
Merge branch 'master' into TriggerMentionOnWhitespaceRemovalRebased

Commit:
Release 7.1.4

Commit:
Merge pull request #145 from mschouest/master

Release 7.1.4

aniyati committed:
Made `textStateBeforeDeletion` public 

Made `textStateBeforeDeletion` public to be used in plugin

aniyati committed:
Removed doc from .m file

aniyati committed:
Updated logic

Updated logic to fix Korean mention fix

Commit:
Merge branch 'master' into master

Commit:
Merge pull request #141 from dannyighsu/master

Fix calling an optional protocol method without selector check

bkoatz committed:
Release 7.1.5

Commit:
Merge pull request #146 from bkoatz/master

Release 7.1.5

Commit:
Merge branch 'master' into TriggerMentionOnWhitespaceRemovalRebased
